### DISCUSSION TITLE
ecr requires parallelMap

### DISCUSSION THREAD

**Discussion by SudoWodo** [SudoWodo/StochOptim#21] [Q&A]
while working on implementing `ecr` I found that It depends on `parallelMap` to evaluate `fitness.fun` and I get the following error if I don't load ecr by `require("ecr", quietly = TRUE)`.
```
Error in parallelMap(function(x) do.call(fitness.fun, c(list(x), list(...))), : 
Level 'ecr.evaluateFitness' not registered
```

 I have tried `requireNamespace("ecr", quietly = TRUE)`  and also added `ecr::` where it was required but that does't work.

What can I do to not use `require()`. Is there some way that I can load  `parallelMap()` without loading entire `ecr`

**Comment by SudoWodo** [comment]
Should we remove ecr? it has some dependencies which I can't manage and have to be stated explicitly with `Imports: parallelMap` in the description file !

**Comment by nashjc** [comment]
Does it help to run a function like optimx::checksolver()?

checksolver <- function(method, allmeth, allpkg){
#    basestats <- c("Nelder-Mead","BFGS","L-BFGS-B","CG","SANN", "nlm", "nlminb", "hjn")
#     cat("method = ",method,"\n")
#     cat("allmeth:"); print(allmeth)
#     cat("allpkg:");print(allpkg)
#    if (method %in% basestats) return(method)
#    Checks if method is available in allmeth
    imeth <- which(method == allmeth)
    if (length(imeth) < 1) {
       warning("Package ",method," not found")
       return(NULL)
    } else {
      pkg <- allpkg[imeth][[1]]
      if ( requireNamespace(pkg, quietly = TRUE)) {
        return(method)
      } else { 
        warning("Package ",pkg," for method ",method," is not available")
        return(NULL)
      }
    }     
    NULL # just in case
}

**Reply by SudoWodo** [reply]
I am running `checkinstallation()`  [check installation] and tried loading the required namespace but it didn't help. I specially noted the commit where I noticed this issue. This is the commit hash : eff3e686ae33b62b1ae2a90883bec42fdb5c23be dated Jul 5, 2022  name of the commit : [a unique ecr error to see run inst/ecr_helper_file](https://github.com/SudoWodo/StochOptim/commit/eff3e686ae33b62b1ae2a90883bec42fdb5c23be)

**Comment by nashjc** [comment]
And I've no idea why the crazy font sizes.

**Reply by SudoWodo** [reply]
Hash (#)  increases the font size.

**Comment by nashjc** [comment]
installing ecr allowed stochoptim to check clean on Linux

**Reply by SudoWodo** [reply]
that's great but it still requires us to load the complete package !

**Comment by nashjc** [comment]
Does it load on "library(stochoptim)" or can it be conditional?

**Reply by SudoWodo** [reply]
it doesn't load on `library(stochoptim)` only when method ecr is used

**Comment by hwborchers** [comment]
How come you decided to include 'ecr' as an optimizer in the StochOptim package? 'ecr' is a "Framework for building evolutionary algorithms ..." and not a ready-to-use stochastic optimizer as those in other packages. Can we please stick to those optimizers we all have agreed to include: DEoptim, DEoptimR, GenSA, PSO (and maybe simpleDE)!

**Reply by SudoWodo** [reply]
actually I noted ecr along with other optimizers on our very first meeting, prof. Cortez mentioned ecr that's why I included it. Now that it is mentioned that ecr is meant to be a "Framework for building evolutionary algorithms ..." It makes sense as to why it had so many unusual methods. Thank you for pointing it out.

**Reply by SudoWodo** [reply]
and yes simpleDE is also included.
